src/github/client/github-client.ts

```diff
@@ -0,0 +1,93 @@
+import { Octokit } from "@octokit/rest";
+import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
+import AppTokenCache from "./app-token-cache";
+import InstallationTokenCache from "./installation-token-cache";
+import AuthToken from "./auth-token";
+
+/**
+ * A GitHub client that supports authentication as a GitHub app.
+ *
+ * @see https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps
+ */
+export default class GitHubClient {
+
+	private readonly axios: AxiosInstance;
+	private readonly appTokenCache: AppTokenCache;
+	private readonly installationTokenCache: InstallationTokenCache;
+	private readonly githubInstallationId: number;
+
+	constructor(
+		appTokenCache: AppTokenCache,
+		installationTokenCache: InstallationTokenCache,
+		githubInstallationId: number,
+		baseURL = "https://api.github.com"
+	) {
+		this.axios = axios.create({
+			baseURL
+		});
+		this.appTokenCache = appTokenCache;
+		this.installationTokenCache = installationTokenCache;
+		this.githubInstallationId = githubInstallationId;
+	}
+
+	/**
+	 * Use this config in a request to authenticate with the app token.
+	 */
+	private asAppConfig(): Partial<AxiosRequestConfig> {
+		const appToken = this.appTokenCache.getAppToken();
+		return {
+			headers: {
+				Accept: "application/vnd.github.v3+json",
+				Authorization: `Bearer ${appToken.token}`
+			}
+		}
+	}
+
+	/**
+	 * Use this config in a request to authenticate with an installation token for the githubInstallationId.
+	 */
+	private async asInstallationConfig(): Promise<Partial<AxiosRequestConfig>> {
+		const installationToken = await this.installationTokenCache.getInstallationToken(
+			this.githubInstallationId,
+			() => this.createInstallationToken(this.githubInstallationId));
+		return {
+			headers: {
+				Accept: "application/vnd.github.v3+json",
+				Authorization: `Bearer ${installationToken.token}`
+			}
+		}
+	}
+
+	/**
+	 * Calls the GitHub API in the name of the GitHub app to generate a token that in turn can be used to call the GitHub
+	 * API in the name of an installation of that app (to access the users' data).
+	 */
+	private async createInstallationToken(githubInstallationId: number): Promise<AuthToken> {
+		const response = await this.axios.post<Octokit.AppsCreateInstallationTokenResponse>(`/app/installations/${githubInstallationId}/access_tokens`, {}, {
+			...this.asAppConfig()
+		});
```

Comment:
This seems weird to me.  Isn't the whole point of the Github Client is to have a centralized Axios instance with the defaults already set so we don't have to think about it?

Reply:
gotcha, makes sense.